### PR TITLE
fix: Canvas multi series checks if series decorate exists before invoking

### PR DESCRIPTION
**Pull request by Ro4052**
Unlike the SVG implementation, `seriesCanvasMulti` invokes `decorate` on individual series. This is breaking in the cases where series does not have a `decorate` function. `seriesCanvasMulti` now checks if the series has a `decorate` before trying to call it.

### CODE REVIEW

File: packages/d3fc-series/src/canvas/multi.js [hunk 1]
```diff
@@ -18,15 +18,19 @@ export default () => {
...
-                adaptedDecorate(c, d, i);
-            });
-
-            dataSeries(seriesData);
-
-            dataSeries.decorate(adaptedDecorate);
+            if (dataSeries.decorate) {
+                const adaptedDecorate = dataSeries.decorate();
+                dataSeries.decorate((c, d, i) => {
+                    base.decorate()(c, data, index);
```
Comment (chrisprice) [inline packages/d3fc-series/src/canvas/multi.js:25]: I'm thinking we should still call `base.decorate` in the case that `dataSeries.decorate` is not defined. What do you think?

File: packages/d3fc-series/src/canvas/multi.js [hunk 2]
```diff
@@ -18,15 +18,19 @@ export default () => {
...
+                    adaptedDecorate(c, d, i);
+                });
+
+                dataSeries(seriesData);
+
+                dataSeries.decorate(adaptedDecorate);
+            } else {
+                dataSeries(seriesData);
```
Comment (chrisprice) [inline packages/d3fc-series/src/canvas/multi.js]: Minor stylistic nit pic - I'm not a fan of duplicating this line.

File: packages/d3fc-series/src/canvas/multi.js (hunk 1 above)
Comment (Ro4052) [reply packages/d3fc-series/src/canvas/multi.js:25]: I don't see any issues with this

File: packages/d3fc-series/src/canvas/multi.js (hunk 2 above)
Comment (Ro4052) [reply packages/d3fc-series/src/canvas/multi.js]: I agree with this point but I wasn't sure how to refactor. I'll take another look.

**Comment by Ro4052** [comment]
@chrisprice, I've made a couple of changes based on your comments. Let me know if there's anything else you think needs doing.

File: packages/d3fc-series/src/canvas/multi.js [hunk 3]
```diff
@@ -18,15 +18,21 @@ export default () => {
...
+            if (dataSeries.decorate) {
+                adaptedDecorate = dataSeries.decorate();
+                dataSeries.decorate((c, d, i) => {
+                    base.decorate()(c, data, index);
+                    adaptedDecorate(c, d, i);
+                });
+            }
 
             dataSeries(seriesData);
+            base.decorate()(context, data, index);
```
Comment (chrisprice) [inline packages/d3fc-series/src/canvas/multi.js]: Sorry it's taken a little while to look at this. It all looks good to me apart from this line, I'm not sure why it's there?

Comment (Ro4052) [reply packages/d3fc-series/src/canvas/multi.js]: That was in reference to your comment about calling `base.decorate` even if `dataSeries.decorate` is undefined. Perhaps I misunderstood? I can remove the line if you'd prefer.

**Comment by chrisprice** [comment]
Would this implementation mean we'd double call it when dataSeries.decorate
is defined?

**Comment by Ro4052** [comment]
Ah, OK, I see the problem. I'll get that sorted.

**Review by chrisprice — APPROVED** [review]
LGTM 👍

**Pull request merged**